### ISSUE TITLE
Forward/backward hooks for C++ torch::nn modules

### ISSUE THREAD

**Issue by yf225** [pytorch/pytorch#25888]
## 🚀 Feature
As part of the Python/C++ API parity work, we would like to provide forward/backward hooks for C++ `torch::nn` modules:
(left: Python API, right: C++ API)
- `torch.nn.Module.register_forward_pre_hook` -> `torch::nn::Module::register_forward_pre_hook`
- `torch.nn.Module.register_forward_hook` -> `torch::nn::Module::register_forward_hook`
- `torch.nn.Module.register_backward_hook` -> `torch::nn::Module::register_backward_hook`

## Possible Steps
- Implement `torch::utils::hooks::RemovableHandle` in C++ API, which mirrors `torch.utils.hooks.RemovableHandle` in Python API.
- Implement `register_forward_pre_hook`, `register_forward_hook` and `register_backward_hook` methods for `torch::nn::Module`.
    - We might need to add `_backward_hooks`, `_forward_pre_hooks` and `_forward_hooks` to `torch::nn::Module` (torch/csrc/api/include/torch/nn/module.h).
- Add appropriate C++ tests for the new API, similar to Python tests.

**Comment by yf225** [comment]
cc. @pbelevich

**Comment by albanD** [comment]
Hi,

You might want to wait for the `backward_hook`s for the python version to be fixed and then only implement the equivalent of the fixed version in cpp?

**Comment by yf225** [comment]
@albanD That sounds great. Do you know if there is a tracking issue for fixing the backward hooks for the Python version?

**Comment by ezyang** [comment]
Note that @malvika2147 implemented hooks on variables.

**Comment by albanD** [comment]
@yf225 I would say https://github.com/pytorch/pytorch/issues/598 is the base issue for backward hooks on nn.Modules.

**Comment by ssbotelh** [comment]
The authors of this paper (https://arxiv.org/abs/1706.02677) claim to achieve great scaling on ResNet50 training by, among other things, overlapping gradient computation and communication: 

> "_as soon as the gradient for a layer is computed, it is aggregated across workers, while gradient computation for the next layer continues_". 

If I understand it correctly, this would only be possible using something like `register_backward_hook` on modules. Since it seems like it'll be a while before this feature is fixed in Python (#598 and #12331) and makes its way to _libtorch_, can anyone think of an alternative way of being notified when the gradient computation is done for a module? Thanks in advance.

**Comment by albanD** [comment]
Hi,

I don't really see why the hooks on Module could help you do that.
But both DataParallel and DistributedDataParallel already overlap computation and communication if I remember correctly.

**Comment by albanD** [comment]
Hi,

The same thing can be done using hooks on Tensors with `register_hook()`. These work as expected.

**Comment by albanD** [comment]
The PR for it is here: https://github.com/pytorch/pytorch/pull/26568

**Comment by yf225** [comment]
@albanD Curious is `register_forward_pre_hook` also blocked by the need to fix `backward_hooks`, or are they independent of each other? Thanks!

**Comment by albanD** [comment]
Hi,

No they are independent.
Both forward_hook and forward_pre_hook are easy to do and not problematic :)

**Comment by ly1231cn** [comment]
@albanD Is register_forward_pre_hook available in the nightly build?

**Comment by albanD** [comment]
I don't think it has been added to the cpp api yet (only available in python). cc @yf225 ?

**Comment by nuka137** [comment]
Is this feature discarded?
I need `register_forward_hook` in order to output model summary (number of parameter, tensor shape, etc...).
It helps us to debug model easily.
There is already [Python version](https://github.com/sksq96/pytorch-summary/blob/4ee5ac5c0787658c7b49f1b8d222757a64c22470/torchsummary/torchsummary.py#L52) which uses `torch.nn.Module.register_forward_hook`.
I think similar API is also needed in C++.

**Comment by albanD** [comment]
It is not. We did not had time to get around to do it.
We would be very happy to help prepare and accept a PR adding this feature though !

**Comment by nuka137** [comment]
@albanD 

Thanks for your reply.
I'm interested in adding `register_forward_hook` .
BTW, `register_backward_hook` seems to be blocked by https://github.com/pytorch/pytorch/issues/25888#issuecomment-529953083 .
So, is it acceptable if I implement only `register_forward_hook` ?

**Comment by albanD** [comment]
Yes only forward hook and pre_forward hook would be sufficient.
More work needs to be done for the backward hooks I'm afraid.

**Comment by nuka137** [comment]
From works for a week, I found it is difficult to expose hook function as simple as Python.

I think we need to add hook function call to [this method](https://github.com/pytorch/pytorch/blob/a371652bc85211a3967672bd0db5ece80f2ebae5/torch/csrc/api/include/torch/nn/pimpl.h#L118-L126) .
But inputs/outputs of `forward` function are not consistent, so we need to expose API arguments explicitly (or use template?).
(ex. [AMSOutput](https://github.com/pytorch/pytorch/blob/e95657b87e9c658c1c3a173ffbc6ca2c08849875/torch/csrc/api/include/torch/nn/modules/adaptive.h) which is an output of `AdaptiveLogSoftmaxWithLoss` while `Tensor` which is an output of `DropoutImpl` and so on)

This hook function should be similar to Python hook function, but I could not have good idea to realize this.
I'm appreciate to give me an advice about it.

**Comment by albanD** [comment]
My guess is that some template magic will be needed here yes.
The main features that are needed here are:
- Call the user hooks before and after the forward call. One takes the `Args...` as input and the other one takes `torch::detail::return_type_of_forward_t<Contained, Args...>`.
- Allow the user to change the values in these hooks. This one I'm not sure how to do.

cc @yf225 do you have an idea of how this can be done?

**Comment by yf225** [comment]
I had some draft PR that might be able to do it: https://github.com/pytorch/pytorch/pull/30484. Haven't got it to the finish line, but maybe can serve as inspiration :) also cc. @glaringlee

**Comment by glaringlee** [comment]
@yf255 Thanks, I'll take a look at it and continue working on it.
@nuka137
I will prioritize this and hope we can finish this pre fwd hook later this year. We currently prioritize the work for adding transformer layer into pytorch c++ apis.

**Comment by nuka137** [comment]
@glaringlee 

Thanks! I hope this feature to land in future.

**Comment by FlyinTeller** [comment]
@glaringlee @yf225  Is there any plan to finish this functionality in the near future? I would much like to use spectral normalization from the C++ API, but this is depending on `register_forward_pre_hook` to be available

**Comment by bjaeger1** [comment]
is there any progress on the implementation of `register_forward_hook` in the C++ API? I would need it to extract feature maps of certain layers of a trained model.

**Comment by spiegelball** [comment]
I am also very curious about this.

**Comment by ezyang** [comment]
There are currently no plans to add this from core team. If someone can come up with a short PR that adds the feature I can help review it, but it may not be so easy to play the same tricks you can do in Python in C++

**Comment by Tihaw** [comment]
Does backward hooks supported by Torchscript module by now?

**Comment by hanjie22** [comment]
Is Module_hooks avaliable now?
